usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -908,9 +918,13 @@ zil_lwb_write_init(zilog_t *zilog, lwb_t *lwb)
 		    ZIO_FLAG_CANFAIL);
 	}
 	if (lwb->lwb_zio == NULL) {
+		if (zilog->zl_cur_used <= zil_slog_limit || !lwb->lwb_slog)
+			prio = ZIO_PRIORITY_SYNC_WRITE;
+		else
+			prio = ZIO_PRIORITY_ASYNC_WRITE;
```

Comment:
I understand your worry about possible starvation, but not sure how realistic it is for SLOG.  If somebody have so high permanent overload on SLOG, he should probably consider upgrading hardware or switching to throughput mode.  Another alternative of writing log to main pool is not exactly perfect from both latency and throughput points of view.